src/components/structures/RoomView.js

```diff
@@ -698,7 +698,44 @@ module.exports = React.createClass({
 
     onJoinButtonClicked: function(ev) {
         var self = this;
-        MatrixClientPeg.get().joinRoom(this.props.roomId).done(function() {
+
+        var cli = MatrixClientPeg.get();
+        var join_defer;
+        var join_promise;
+        // if this is the first room we're joining, checkthe user has a display name
+        // and if they don't, prompt them to set one.
+        // NB. This unfortunately does not re-use the ChangeDisplayName component because
+        // it doesn't behave quite as desired here (we want an input field here rather than
+        // content-editable, and we want a default).
+        if (MatrixClientPeg.get().getRooms().length == 0) {
+            join_promise = cli.getProfileInfo(cli.credentials.userId).then((result) => {
+                if (!result.displayname) {
+                    var SetDisplayNameDialog = sdk.getComponent('views.dialogs.SetDisplayNameDialog');
+                    var dialog_defer = q.defer();
+                    var dialog_ref;
+                    var modal;
+                    var dialog_instance = <SetDisplayNameDialog currentDisplayName={result.displayname} ref={(r) => {
+                            dialog_ref = r;
+                    }} onFinished={() => {
+                        var new_displayname = dialog_ref.getValue() || dialog_ref.getDefaultValue();
+                        cli.setDisplayName(new_displayname).done(() => {
+                            dialog_defer.resolve();
+                        });
+                        modal.close();
+                    }} />
+                    modal = Modal.createDialogWithElement(dialog_instance);
+                    return dialog_defer.promise;
+                }
+            });
+        } else {
+            join_defer = q.defer();
+            join_promise = join_defer.promise;
+            join_defer.resolve();
```

Comment:
`join_defer` is pointless here. From the looks of it, all we're doing is making a resolved promise if they have > 0 rooms which we can then use on `:736`. I'd delete this entire `else` block and just make a resolved promise on `:704` which is then clobbered by the `if` statement if required. Also, the fact that `join_promise` isn't a promise for `/join` is mighty misleading. I would call it `display_name_promise` or something. So overall, you'd have something like:

```javascript
var display_name_promise = q(); // already resolved
if (rooms == 0) {
  display_name_promise = cli.getProfileInfo.then(/* dialog bit if required */)
}
display_name_promise.then(() => {
  return joinRoom();
}).done(...)
```